### PR TITLE
Preserve typed SSSOM set comments

### LINKED ISSUE #133: sssom: parse_tsv/serialize silently drops set-level trailer comments (# #…)
## Summary

purrdf's SSSOM codec **silently drops set-level trailer comments** on the round trip. `parse_tsv` treats a `# #…` line as a documentation comment and does not surface it as mapping-set metadata, and the serializer does not re-emit it. Any set-level provenance a producer folds into the trailer (refused-/deferred-mapping declarations, lossy-projection notes, curation notes) is lost.

## Impact

For a byte-preserving / lossless SSSOM surface this is a data-loss bug, not a cosmetic one: the trailer is the only place SSSOM carries set-scoped prose that is not a per-row `mapping_*` column, and the codec makes it unrepresentable end-to-end.

## Requested (maximal)

1. **Round-trip trailer comments losslessly.** `parse_tsv` should capture trailing `#`/`# #` comment blocks into a structured field on the parsed mapping set (e.g. `set.trailer_comments: Vec<String>` distinct from the header `#`-prefixed YAML metadata block), and the serializer should re-emit them verbatim in the same position.
2. Distinguish the **header metadata block** (leading `#` YAML) from **trailer prose** (trailing `# #`) so neither clobbers the other.
3. Offer a typed way to attach set-level provenance so callers do not have to hand-splice comment lines at all — e.g. a first-class `mapping_set_provenance` / `comment` slot that serializes to the trailer.

## Current downstream workaround (retire once fixed)

gmeow serializes via purrdf, then **splices** its provenance trailer back into the emitted TSV as `# #…` lines after the last data row (a `splice_trailer` post-pass), specifically because `parse_tsv` skips them and the serializer will not emit them. This exists only to keep the SSSOM output non-lossy and can be deleted once the codec round-trips trailers natively.

### PR DESCRIPTION
**Pull request by paudley**
## Summary

- model SSSOM set-level ordinary comments and provenance as validated, lossless document-envelope values;
- distinguish YAML-like header metadata, standard-compatible before-table provenance, interleaved row comments, and the preserved after-table comment extension;
- retain parsed TSV column declarations, including extension columns on zero-row mapping sets;
- serialize comments, layouts, mappings, and line endings deterministically through the existing Rust and umbrella facades.

## Requirement mapping

| Requirement | Evidence |
| --- | --- |
| Capture and re-emit post-table `#` / `# #` blocks | `SssomSetComment` with `AfterTable`; exact-output and fixed-point tests |
| Preserve GMEOW-style provenance before the table | `BeforeTable` parsing/serialization and typed `append_provenance` default |
| Keep metadata distinct from prose | parser classification tests, including metadata-looking trailer prose and an open `curie_map` |
| Avoid text splicing | fallible raw/ordinary/provenance constructors and atomic mapping-set append APIs |
| Preserve exact comment content and deterministic bytes | raw Unicode line retention, trailing-space coverage, CRLF-to-LF canonicalization, duplicate-order coverage |
| Preserve zero-row schemas | validated retained `SssomColumnLayout` and zero-row extension-column fixed point |
| Keep semantic behavior unchanged | validation-diagnostic and RDF-quad invariance tests |
| Expose one canonical public behavior | exports through `purrdf-core`, `purrdf-rdf`, and a compiling `purrdf` facade test |

## Validation

- `cargo test -p purrdf-core sssom --locked` (26 passed)
- `cargo test -p purrdf-rdf --test never_panic sssom --locked`
- `cargo test -p purrdf facade_exposes_typed_sssom_document_envelope --locked`
- `cargo clippy -p purrdf-core -p purrdf-rdf -p purrdf --all-targets --locked -- -D warnings`
- `make check` (full workspace, conformance, hygiene/generated checks, and wasm build)
- mechanical incomplete-work and whitespace scans

## Compatibility and residual risk

Adding the envelope and retained-layout fields changes external `SssomMappingSet` struct literals. `SssomMappingSet::new(meta, mappings)` is the stable construction path and preserves the prior inferred-column behavior. The existing parser, validator, RDF projection, mapping sort, and unknown-field behavior remain covered. No vocabulary IRI, dependency, Cargo feature, generated artifact, or emitted RDF identity was added.

Closes #133

## Summary by CodeRabbit

* **New Features**
  * Added support for preserving and round-tripping SSSOM document-level comments, provenance, and after-table envelope content.
  * Preserved declared column layouts, including empty datasets, while deterministically handling new fields during serialization.
  * Expanded access to SSSOM types and validation constants.

* **Documentation**
  * Documented lossless SSSOM envelope handling and deterministic line-ending behavior.

* **Tests & Performance**
  * Added coverage for comment validation, placement, layout preservation, and serialization.
  * Added benchmarks for layout validation and sparse SSSOM serialization.

### CODE REVIEW

**Comment by paudley** [comment]
## Issue summary

The native SSSOM codec currently discards set-scoped `#` and `# #` comment blocks. The correction will make those blocks first-class document-envelope data, preserve the existing distinction between YAML-like header metadata and prose provenance, and let Rust callers attach validated set-level comments without manipulating serialized text. The model will preserve both the pre-table YAML-comment placement used by current GMEOW files and the reported post-table trailer extension. Typed provenance construction will default to the standard-compatible pre-table placement because the authoritative SSSOM/TSV structure requires the mappings block to follow the metadata block immediately. The same model correction will retain the declared TSV column layout so a zero-row set cannot lose optional or extension columns during serialization.

## Baseline defaults

- Review adversarially and prove every stated requirement.
- Deliver the complete bug fix now and prove every completeness-contract row.
- Keep repository documentation free of GitHub process references.
- Fix any in-scope bug encountered while implementing this codec correction.

## Standing constraints

- Use a Rust-first, typed implementation and expose the same canonical behavior through existing public facades.
- Preserve maximal information, deterministic bytes, WebAssembly portability, and hard validation failures.
- Add no Cargo feature, optional dependency, filesystem requirement, runtime randomness, ontology IRI, or generated-file hand edit.
- Keep RDF 1.2 and RDF-star carrier behavior unchanged; SSSOM comments remain projection-envelope data rather than RDF identity.
- `CONSTITUTION.md`: None.

## Completeness contract

| Requirement or acceptance criterion | Delivery task |
| --- | --- |
| Capture post-table `#` and `# #` comment blocks as structured mapping-set data | Tasks 1 and 2 |
| Re-emit every captured comment line verbatim, in input order, after the table | Tasks 1 and 3 |
| Preserve GMEOW-style `# #` provenance blocks between header metadata and the TSV column header | Tasks 1, 2, and 3 |
| Keep leading YAML-like metadata independent from prose comments so `meta.comment`, extra scalars, and `curie_map` are never clobbered | Tasks 2 and 3 |
| Provide typed caller APIs for ordinary set comments and mapping-set provenance without text splicing, defaulting new provenance to standard-compatible pre-table placement | Task 1 |
| Reject missing markers, embedded CR/LF, ambiguous single-hash before-table comments, and kind/placement mismatches through a public typed error | Tasks 1 and 4 |
| Preserve deterministic parse/serialize fixed points for before-table and after-table blocks, including an empty `# #` provenance line | Tasks 2, 3, and 4 |
| Preserve the declared TSV header schema for zero-row sets, including unknown extension columns | Tasks 1, 2, 3, and 4 |
| Keep mapping parsing, true source-line diagnostics, validation, RDF projection, and unknown header/column preservation intact | Tasks 2, 3, and 4 |
| Export the comment, placement, kind, error, and column-layout model consistently from `purrdf-core`, `purrdf-rdf`, and the umbrella facade | Tasks 1 and 3 |
| Keep comment attachment semantically inert for validation and RDF projection | Tasks 3 and 4 |
| Complete isolated branch setup, per-task commits and pushes, issue updates, full validation, and a non-draft PR | Tasks 0 through 5 |

## Enhancements considered and declined

- Adopted transformation: model comments as a lossless SSSOM document envelope around the canonical mapping table, using `SssomSetComment` values with explicit before-table and after-table placement plus computed ordinary/provenance kind. This subsumes both the reported trailer case and the established GMEOW representation instead of encoding one producer's physical layout as a special case.
- Adopted transformation: retain the parsed TSV column layout as a validated typed value. Parsed zero-row sets then preserve their declared schema, while programmatically constructed sets continue to infer the canonical layout.
- Adopted leverage: reuse the codec's existing raw-line scan and canonical serializer; no new dependency or parsing layer is needed.
- Adopted leverage: use one shared comment-line classifier for constructors, leading-region parsing, suffix parsing, and kind inspection.
- Adopted utility: add validated ordinary-comment and provenance constructors plus atomic mapping-set append methods, so callers work with typed values rather than comment prefixes; new provenance uses the standard-compatible pre-table placement unless a caller explicitly requests preservation of an after-table extension.
- Adopted robustness: preserve exact Unicode line content excluding its physical line terminator, canonicalize output to LF, and cover CRLF input, Unicode, trailing spaces, zero-row sets, duplicate comments, both placements together, blank provenance content, metadata separation, fixed-point serialization, validation/RDF invariance, and source-line diagnostics.
- Arbitrary row-anchored interleaved comments are not adopted because they are row-local rather than set-level and conflict with the established canonical row sort; retaining them as set provenance would change their meaning.
- An RDF vocabulary projection for prose provenance is not adopted because PurRDF cannot invent or assume vocabulary IRIs. The typed projection-envelope value is lossless and caller-neutral.

## Task 0: isolated worktree and branch

1. Post this reviewed, mechanically clean plan to issue #133 before repository mutation.
2. Verify the git-common exclude file contains the exact `/.worktrees/` rule.
3. Fetch `origin/main` and create `.worktrees/133-sssom-set-comments` on `paudley/133-sssom-set-comments` from the current remote head.
4. Confirm the worktree branch is not `main`, the main checkout remains untouched, and the new worktree starts clean.
5. Run the focused SSSOM baseline test before edits.

## Task 1: typed set-level comment model

1. Add `SssomCommentPlacement::{BeforeTable, AfterTable}`, `SssomCommentKind::{Ordinary, Provenance}`, and a `SssomSetComment` value with private raw content and public placement, kind, and raw-line accessors.
2. Add a public `SssomCommentError` with explicit missing-marker, embedded-line-terminator, ambiguous-before-table, and kind-mismatch variants. Before-table values accept only an unambiguous second-marker provenance line; after-table values accept ordinary and provenance lines.
3. Add fallible raw, ordinary, and provenance constructors. Ordinary content whose trimmed prefix begins with `#` hard-fails instead of changing kind. Preserve the empty `# #` form.
4. Add a validated `SssomColumnLayout` that rejects empty, duplicate, tabbed, or multiline column names and retains parsed header order.
5. Add `set_comments` and the column layout to `SssomMappingSet`, plus a stable `SssomMappingSet::new(meta, mappings)` constructor and atomic append APIs for pre-table provenance, post-table ordinary comments, and post-table provenance. Programmatic construction starts with inferred canonical columns.
6. Re-export every new public type through the existing public crate surfaces.
7. Run focused unit tests, commit as one coherent model/API change, push, and post the validation and commit hash to the issue.

## Task 2: lossless parser classification

1. Parse `# #` lines in the leading metadata region as before-table provenance rather than YAML scalars; reject a single-hash prose line there so it cannot masquerade as metadata.
2. After the TSV header, find the final nonblank, noncomment data line; capture only comment lines after it as after-table comments in source order. For a zero-row set, the suffix starts after the TSV header. Interleaved comments remain body comments and keep their diagnostic-offset behavior.
3. Populate the validated declared-column layout from the TSV header before mapping parsing.
4. Keep ordinary header scalars, `curie_map`, mappings, blank-line handling, and interleaved body-comment diagnostic offsets unchanged.
5. Cover leading and trailing blocks, zero-row input with an extension column, metadata/comment coexistence, metadata-looking post-table comments, CRLF input, significant whitespace, mixed comment kinds, duplicates, and malformed inputs.
6. Run focused tests, commit the parser correction, push, and post the validation and commit hash to the issue.

## Task 3: deterministic serializer and integration wiring

1. Emit before-table provenance after canonical metadata and before the TSV column header; emit after-table comments after the canonical sorted mapping rows. Preserve exact raw line content and stable order within each placement.
2. Use a retained parsed column layout when present and append newly populated columns deterministically; infer the canonical layout for programmatically constructed sets.
3. Update module and book documentation to describe the typed envelope, standard-compatible default, after-table preservation extension, LF normalization, and caller-neutral semantics without repository-process references.
4. Add exact-output and parse/serialize fixed-point tests for bare `#`, empty `# #`, Unicode, trailing spaces, mixed/duplicate comments, both placements, zero-row extension columns, and programmatic constructors.
5. Prove comments leave `validate` diagnostics and `to_rdf` quads unchanged, invalid appends are atomic, metadata fields remain independent, and umbrella-facade construction compiles.
6. Run focused tests, commit the serializer/docs/integration change, push, and post the validation and commit hash to the issue.

## Task 4: validation gates

1. Run focused SSSOM tests and never-panic coverage.
2. Run strict formatting and clippy for every touched crate.
3. Run `make check` for the complete workspace, including hygiene, generated drift, tests, and every release crate on `wasm32-unknown-unknown`.
4. Run the mechanical no-partial-implementation scan against `origin/main...HEAD` and classify every match with evidence.
5. Verify the worktree is clean and every completeness-contract row has direct evidence.

## Task 5: publication

1. Push the final validated head and post the full validation result to the issue.
2. Open a non-draft PR with `Closes #133`, the requirement mapping, implementation summary, commands actually run, and residual-risk statement. Record that adding fields to the public mapping-set value requires existing external struct literals to use the new stable constructor and typed append APIs.
3. Post this reviewed plan to the PR, report the branch and commit list, and begin the required five-minute CI/bot feedback window before stage 2.

**Comment by paudley** [comment]
## Stage 2 gap analysis

PR #143 satisfies every functional requirement in issue #133: both document-envelope placements round-trip, metadata stays independent, typed attachment is available, zero-row layouts survive, emitted bytes are deterministic, and validation/RDF semantics are invariant. Full local validation was green at the reviewed head.

### Actionable gaps

| ID | Priority | Gap | Complete fix | Validation | Commit/update boundary |
| --- | --- | --- | --- | --- | --- |
| G1 | MEDIUM | `SssomColumnLayout::new` allocates a `BTreeSet` and clones every accepted column solely for duplicate detection; `parse_body` then clones the complete header again before constructing the layout. | Use allocation-free linear duplicate detection for the bounded SSSOM column vector, move the parsed header directly into the validated layout, and borrow that layout while parsing rows. | Focused constructor/parser tests, strict clippy, SSSOM benchmark comparison. | One signed performance-remediation commit; push and update issue/PR. |
| G2 | MEDIUM | `serialization_columns` repeatedly scans every mapping for each known column and allocates/clones a second `seen` tree. | Traverse mappings once, record the eight known-column population flags in a stack array, union extension keys deterministically during the same pass, and use direct small-vector membership checks when appending. Preserve exact column order and blank extension-column intent. | Existing exact-output/fixed-point tests plus a sparse multi-row known/extension union regression, strict clippy, SSSOM benchmark comparison. | Same coherent performance-remediation commit; push and update issue/PR. |
| G3 | MEDIUM | The review requests codec hot-path performance changes, but no SSSOM Criterion harness exists to measure them as required by repository performance policy. | Add a report-only benchmark for validated 32-column layout construction and sparse 4k-row deterministic serialization. Capture the current head as the baseline before production edits and compare the remediated implementation against it. | `cargo bench -p purrdf-core --bench sssom -- --save-baseline stage2-before`, then `cargo bench -p purrdf-core --bench sssom -- --baseline stage2-before`. | Benchmark ships with the same coherent commit and its evidence is posted with the remediation update. |

### Non-actionable findings

- CodeRabbit reported an organization usage cap and supplied no code finding; this is informational.
- CI jobs still running are processing state, not implementation gaps. Completed CodeQL, docs, MSRV, projection, wasm, workspace, and other reported jobs are green.
- The issue body's phrase `deferred-mapping declarations` names provenance data that must be preserved; it is not implementation deferral.
- The validation comment's TODO/FIXME wording reports a clean scan; it does not add incomplete work.
- No unresolved functional thread, requested change, stale generated artifact, feature flag, vocabulary dependency, or documentation process reference was found.

### Execution order

1. Add the report-only benchmark and measure the unchanged reviewed implementation.
2. Apply G1 and G2 without changing serializer output or public semantics; add the sparse-union regression.
3. Re-run the benchmark against the saved baseline, focused tests, never-panic tests, strict workspace formatting/clippy, and `make check`.
4. Commit and push the single coherent remediation, update issue and PR with measured/validation evidence, then re-fetch all comments, reviews, thread states, and checks.
5. Resolve the two threads only after the fixes are present and verified; publish the final Stage 2 readiness summary.

**Comment by paudley** [comment]
### Stage 2 remediation: G1–G3 fixed

Commit `56db1a1` (`perf(sssom): streamline column selection`) completes the three medium-priority performance gaps:

- `SssomColumnLayout::new` now detects duplicates directly in the bounded retained vector, eliminating the duplicate-detection tree and per-column clone.
- `parse_body` moves the parsed header into the validated layout and borrows the retained columns for every row.
- `serialization_columns` now traverses mappings once, records the eight known-column population flags on the stack, unions extension keys during that pass, and avoids the cloned `seen` tree.
- A report-only Criterion harness covers validated 32-column layouts and deterministic sparse 4,096-row serialization; the sparse union regression proves known and blank/nonblank extension columns retain exact canonical ordering.

Validation:

- `cargo test -p purrdf-core sssom --locked`: 27 passed.
- `cargo clippy -p purrdf-core -p purrdf-rdf -p purrdf --all-targets --locked -- -D warnings`: passed.
- `cargo fmt --all -- --check`: passed.
- whitespace check: passed.
- `cargo bench -p purrdf-core --bench sssom --locked -- --baseline stage2-before`: both measurements improved significantly:
  - validated 32-column layout: 992.76 ns midpoint, 53.79% faster;
  - sparse 4,096-row serialization: 455.63 µs midpoint, 45.35% faster.

The commit is signed and pushed. I am re-running the full branch gates and will resolve the two Gemini threads after the complete validation evidence is green.

File: crates/rdf-core/src/sssom.rs [hunk 1]
```diff
@@ -159,13 +162,342 @@ pub struct SssomMapping {
...
+            }
+            if column.contains(['\t', '\r', '\n']) {
+                return Err(SssomColumnLayoutError::InvalidDelimiter(column));
+            }
+            if !seen.insert(column.clone()) {
+                return Err(SssomColumnLayoutError::DuplicateName(column));
+            }
+            retained.push(column);
+        }
+        Ok(Self { columns: retained })
```
[the remark this answers is not in the document]
Comment (paudley) [reply crates/rdf-core/src/sssom.rs:406]: Addressed in 56db1a1: duplicate detection now uses the bounded retained vector, and parse_body moves the parsed header into the validated layout. Focused tests, strict Clippy, make check, and the Criterion comparison are green; the 32-column case improved by 53.79%.

File: crates/rdf-core/src/sssom.rs [hunk 2]
```diff
@@ -668,11 +1041,100 @@ pub fn serialize_tsv(set: &SssomMappingSet) -> String {
...
+        .collect();
+    for column in extra_columns {
+        if !seen.contains(column) {
+            let column = column.to_owned();
+            seen.insert(column.clone());
+            columns.push(column);
+        }
+    }
+    columns
+}
```
[the remark this answers is not in the document]
Comment (paudley) [reply crates/rdf-core/src/sssom.rs:1112]: Addressed in 56db1a1: serialization now performs one mapping traversal, keeps eight known-column flags on the stack, unions extension keys in that pass, and uses the small output vector for membership. Exact ordering tests, strict Clippy, make check, and Criterion are green; sparse 4k-row serialization improved by 45.35%.

**Comment by paudley** [comment]
### Stage 2 complete

PR #143 now satisfies the original issue, the standing constraints, and all review feedback.

- G1 fixed: bounded layout duplicate detection no longer allocates a tree or clones accepted columns, and parsed headers move directly into the retained layout.
- G2 fixed: serializer column selection performs one mapping traversal with stack flags and a deterministic extension-key union, without repeated full scans or a cloned membership tree.
- G3 fixed: the new report-only SSSOM Criterion harness measures both affected hot paths; the remediated implementation is 53.79% faster for validated 32-column layouts and 45.35% faster for sparse 4,096-row serialization.
- Reviewer feedback: both Gemini threads were answered with evidence and resolved; the final audit reports 0 unresolved threads. CodeRabbit supplied no actionable finding because of its organization review limit.
- Pushed signed remediation: `56db1a1d6c712ffdf1820b1d2d96c7e32b86cdc8`.

Validation is green:

- `cargo test -p purrdf-core sssom --locked`: 27 passed.
- `cargo test -p purrdf-rdf --test never_panic sssom --locked`: 2 passed.
- `cargo test -p purrdf facade_exposes_typed_sssom_document_envelope --locked`: passed.
- strict touched-crate Clippy, formatting, whitespace, and Criterion baseline comparison: passed.
- `make check`: passed in full, including workspace Clippy/build/tests/doctests, hygiene, generated drift, frozen corpora, version/issue-reference checks, kernel ring-fence, and every required wasm build.
- Latest GitHub correctness/release jobs are green: workspace, test, pytest, conformance, C API, wasm, MSRV, projection oracles, docs/book, analysis, and CodeQL. The only active workflow is the explicitly report-only benchmark run.
- Deferral scan: no code or commit-message hits; the sole discussion hit is the already-classified phrase “deferred-mapping declarations,” which names SSSOM provenance data rather than incomplete implementation.
- Branch is clean and local/remote heads match.

No implementation gap or blocker remains. PR #143 is ready for Stage 3 finalization.

**Comment by paudley** [comment]
sssom: preserve typed set-level document comments

## Summary

Make set-scoped SSSOM comments a validated, lossless document envelope. PurRDF
now distinguishes YAML-like metadata, standard-compatible before-table
provenance, interleaved body comments, and preserved after-table comments while
retaining source column declarations and deterministic serialization.

## Requirements satisfied

- Capture final post-table `#` and `# #` blocks as structured mapping-set data
  and re-emit every line verbatim, in original order and position.
- Preserve before-table `# #` provenance independently from leading metadata,
  including `meta.comment`, extra scalar fields, and open `curie_map` blocks.
- Provide typed ordinary-comment and provenance constructors plus atomic append
  APIs, with standard-compatible before-table placement as the provenance
  default.
- Reject missing markers, embedded line terminators, ambiguous before-table
  ordinary comments, kind mismatches, and invalid TSV column declarations
  through public typed errors.
- Preserve deterministic parse/serialize fixed points for both envelope
  placements, exact Unicode/trailing-space content, empty provenance, CRLF
  input normalized to LF output, duplicate comments, and zero-row sets.
- Retain the declared TSV column schema and order, including unknown extension
  columns, while appending newly populated columns deterministically.
- Keep source-line diagnostics, validation, RDF projection, mapping sorting,
  unknown metadata, and RDF 1.2/RDF-star identity unchanged.
- Export the complete comment and column-layout model through `purrdf-core`,
  `purrdf-rdf`, and the umbrella `purrdf` facade.

## Files and purpose

- `crates/rdf-core/src/sssom.rs`: typed comment/layout model, parser envelope
  classification, deterministic serializer, bounded single-pass column
  selection, public errors, and exhaustive semantic/exact-output tests.
- `crates/rdf-core/src/lib.rs`: public core exports for the new SSSOM types.
- `crates/rdf-core/Cargo.toml` and `crates/rdf-core/benches/sssom.rs`: report-only
  Criterion coverage for validated layouts and sparse deterministic
  serialization.
- `crates/rdf/src/lib.rs`: re-export the canonical SSSOM model from the RDF
  surface.
- `crates/purrdf/src/lib.rs`: umbrella-facade exports and construction proof.
- `docs/book/src/slices.md`: user-facing document-envelope, placement, typed API,
  and LF-normalization behavior.
- No files were deleted and no generated projection was hand-edited.

## Validation

- `cargo test -p purrdf-core sssom --locked`: 27 passed.
- `cargo test -p purrdf-rdf --test never_panic sssom --locked`: both arbitrary
  byte and structured-input property tests passed.
- `cargo test -p purrdf facade_exposes_typed_sssom_document_envelope --locked`:
  passed before and after synchronizing `origin/main`.
- `cargo clippy -p purrdf-core -p purrdf-rdf -p purrdf --all-targets --locked -- -D warnings`:
  passed before and after synchronization.
- `cargo bench -p purrdf-core --bench sssom --locked -- --baseline
  stage2-before`: validated 32-column layouts improved by 53.79%; sparse 4,096-row
  serialization improved by 45.35%.
- `make check`: passed in full before synchronization, including workspace
  formatting/Clippy/build/tests/doctests, hygiene, generated drift, frozen
  corpora, version/issue-reference checks, kernel ring-fence, and all required
  wasm builds.
- The synchronized GitHub correctness/release run passed workspace, full tests,
  Python, conformance, C API, wasm, MSRV, projection oracles, docs/book,
  analysis, and CodeQL. The independent benchmark workflow remained
  report-only; CodeRabbit was rate-limited and added no review thread.
- Whitespace, commit-signature, remote-head, review-thread, and mechanical
  incomplete-work audits passed.

## Standing constraints

The branch satisfies `.goals`: its canonical implementation is Rust-first and
binding-neutral; it preserves maximum information and deterministic bytes; it
adds no Cargo feature, optional dependency, runtime randomness, filesystem
requirement, or vocabulary IRI; and it remains wasm-portable without changing
RDF 1.2/RDF-star carrier identity.

## Deferral Gate

Passed. The branch diff and commit messages have no incomplete implementation
markers. The discussion scan finds only “deferred-mapping declarations,” which
is the issue's name for provenance text that the codec must preserve. It does
not describe unfinished code, and the complete parse/serialize/API behavior is
covered by exact-output and fixed-point tests.

## Implementation notes

- `SssomSetComment` retains the exact logical line without its physical line
  terminator and computes ordinary/provenance kind from the validated marker.
- Before-table ordinary comments hard-fail because they are ambiguous with the
  YAML-like metadata region; before-table provenance is unambiguous, while both
  kinds remain losslessly representable after the table.
- Parsed layouts stay exact even with zero rows. Programmatic mapping sets infer
  canonical known columns and lexically ordered extension columns.
- Serializer column selection traverses mappings once, records eight known
  column flags on the stack, and unions extension keys deterministically.
- Set comments are document-envelope data: tests prove they do not affect
  validation diagnostics or RDF projection quads.
- `origin/main` merged with the default `ort` strategy without conflicts. No
  manual conflict resolution was required; focused SSSOM, facade, and strict
  Clippy validation passed on the synchronized branch.
